pkg/ip/ip.go

```diff
@@ -0,0 +1,219 @@
+
+	"github.com/asaskevich/govalidator"
+)
+
+const (
+	ipv4BitLen = 8 * net.IPv4len
+	ipv6BitLen = 8 * net.IPv6len
+)
+
+// ByMask is used to sort a list of IP networks by the size of their masks.
+type ByMask []*net.IPNet
+
+func (s ByMask) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s ByMask) Less(i, j int) bool {
+	iPrefixSize, _ := s[i].Mask.Size()
+	jPrefixSize, _ := s[j].Mask.Size()
+	if iPrefixSize == jPrefixSize {
+		byteArrComp := bytes.Compare(s[i].IP, s[j].IP)
+		if byteArrComp < 0 {
+			return true
+		}
+		return false
+	}
+	return iPrefixSize < jPrefixSize
+}
+
+func (s ByMask) Len() int {
+	return len(s)
+}
+
+// RemoveCIDRs removes the specified CIDRs from another set of CIDRs. If a CIDR to remove is not
+// contained within the CIDR, the CIDR to remove is ignored. A slice of CIDRs is
+// returned which contains the set of CIDRs provided minus the set of CIDRs which
+// were removed.
+func RemoveCIDRs(allowCIDRs, removeCIDRs []*net.IPNet) ([]*net.IPNet, error) {
+
+	// Ensure that we iterate through the provided CIDRs in order of largest
+	// subnet first.
+	sort.Sort(ByMask(removeCIDRs))
+
+PreLoop:
+	// Remove CIDRs which are contained within CIDRs that we want to remove;
+	// such CIDRs are redundant.
+	for j, removeCIDR := range removeCIDRs {
+		for i, removeCIDR2 := range removeCIDRs {
+			if i == j {
+				continue
+			}
+			if removeCIDR.Contains(removeCIDR2.IP) {
+				removeCIDRs = append(removeCIDRs[:i], removeCIDRs[i+1:]...)
+				// Re-trigger loop since we have modified the slice we are iterating over.
+				goto PreLoop
+			}
+		}
+	}
+
+Loop:
+	for j, remove := range removeCIDRs {
+		for i, allowCIDR := range allowCIDRs {
+
+			// Don't allow comparison of different address spaces.
+			if !govalidator.IsIPv4(allowCIDR.IP.String()) && govalidator.IsIPv4(remove.IP.String()) ||
+				govalidator.IsIPv4(allowCIDR.IP.String()) && !govalidator.IsIPv4(remove.IP.String()) {
+				return nil, fmt.Errorf("cannot compare different address spaces")
+			}
+
+			// Only remove CIDR if it is contained in the subnet we are allowing.
+			if allowCIDR.Contains(remove.IP) {
+				nets, err := removeCIDR(allowCIDR, remove)
+				if err != nil {
+					return nil, err
+				}
+
+				// Remove CIDR that we have just processed and append new CIDRs
+				// that we computed from removing the CIDR to remove.
+				allowCIDRs = append(allowCIDRs[:i], allowCIDRs[i+1:]...)
+				allowCIDRs = append(allowCIDRs, nets...)
+				goto Loop
+			} else if remove.Contains(allowCIDR.IP.Mask(allowCIDR.Mask)) {
+				// If a CIDR that we want to remove contains a CIDR in the list
+				// that is allowed, then we can just remove the CIDR to allow.
+				allowCIDRs = append(allowCIDRs[:i], allowCIDRs[i+1:]...)
+				goto Loop
+			}
+		}
+		removeCIDRs = append(removeCIDRs[:j], removeCIDRs[j+1:]...)
+		goto Loop
+	}
+
+	return allowCIDRs, nil
+}
+
+func getFirstIP(ipNet *net.IPNet) *net.IP {
+	var mask net.IP
+
+	if govalidator.IsIPv6(ipNet.IP.String()) {
```

Comment:
In the interest of not taking on additional dependencies, can you just use `ipNet.IP.To4() != nil`?

Reply:
Done, and removed the dependency as well.